crates/ide/src/ide/assists/add_to_lambda.rs

```diff
@@ -0,0 +1,84 @@
+//! add an undefined name to the top-level lambda
+use super::{AssistKind, AssistsCtx};
+use crate::def::Name;
+use crate::TextEdit;
+use syntax::ast::{self, AstNode};
+use syntax::{TextRange, TextSize};
+
+pub(super) fn add_to_lambda(ctx: &mut AssistsCtx<'_>) -> Option<()> {
+    let name = ctx.covering_node::<ast::Ref>()?.token()?;
+    let name = name.text();
+
+    // the whole file should be a lambda
+    let pat = ast::Lambda::cast(ctx.ast.syntax().first_child()?)?
+        .param()?
+        .pat()?;
+
+    // name should be undefined
+    if ctx
+        .db
+        .module(ctx.frange.file_id)
+        .names()
+        .any(|(_, Name { text, .. })| text == name)
+    {
+        return None;
+    };
```

Comment:
We should check the name resolution and apply this action to unresolved references, not based on text-wise searching.

Reply:
Can you elaborate on how I should do that?

Reply:
I think the most efficient way might be to bundle all the diagnostics together and go over `ctx.db.module.diagnostics`

Reply:
> Can you elaborate on how I should do that? How do I convert something to an `Idx`

Use [`NameResolution::get`](https://github.com/oxalica/nil/blob/46eeb57f9b7ea1cdd061dc5b40ea7f0146124d76/crates/ide/src/def/nameres.rs#L237) to get the resolution result for an `Expr`. `None` means undefined.
You also need `source_map` to get `Expr` from AST node.

> I think the most efficient way might be to bundle all the diagnostics together and go over ctx.db.module.diagnostics

Generally diagnostics are for human, and not for our logic.